### PR TITLE
Enable WireClient tests

### PR DESCRIPTION
**Pull request by narrieta**
Re-enables the tests disabled by #1799 

I'll annotate the PR with the relevant changes.

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1800)

### CODE REVIEW

File: azurelinuxagent/common/protocol/wire.py [hunk 1]
```diff
@@ -520,11 +520,6 @@ def __init__(self, endpoint):
         logger.info("Wire server endpoint:{0}", endpoint)
         self._endpoint = endpoint
         self._goal_state = None
-        self._hosting_env = None
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/wire.py]: I forgot to remove these on #1777

File: tests/protocol/mock_wire_protocol.py [hunk 2]
```diff
@@ -28,48 +28,56 @@ def create(mock_wire_data_file):
     Creates a mock WireProtocol object that will return the data specified by 'mock_wire_data_file' (which must
     follow the structure of the data files defined in tests/protocol/mockwiredata.py).
 
-    NOTE: This function creates mocks for azurelinuxagent.common.utils.restutil.http_get and
+    NOTE: This function creates mocks for azurelinuxagent.common.utils.restutil.http_request and
```
Comment (narrieta) [inline tests/protocol/mock_wire_protocol.py]: I had to mock 1 level deeper (http_request instead of http_get) to address all the scenarios in the tests,

File: tests/protocol/mock_wire_protocol.py [hunk 3]
```diff
@@ -28,48 +28,56 @@ def create(mock_wire_data_file):
...
-        def mock_http_get(url, *args, **kwargs):
-            if mock_data_re.match(url) is None:
-                return original_http_get(url, *args, **kwargs)
-            return protocol.mock_wire_data.mock_http_get(url, *args, **kwargs)
+        original_http_request = http_request
+
+        def mock_http_request(method, url, data, headers=None, use_proxy=False, max_retry=DEFAULT_RETRIES, retry_codes=RETRY_CODES, retry_delay=DELAY_IN_SECONDS):
+            if method == 'GET' and mock_data_re.match(url) is not None:
+                return protocol.mock_wire_data.mock_http_get(url, headers, use_proxy, max_retry, retry_codes, retry_delay)
+            elif method == 'POST':
```
Comment (narrieta) [inline tests/protocol/mock_wire_protocol.py]: added support for POST

File: tests/protocol/mock_wire_protocol.py [hunk 4]
```diff
@@ -28,48 +28,56 @@ def create(mock_wire_data_file):
...
+    def stop():
+        stop_mock_crypt_util()
+        stop_mock_http_request()
+
     protocol = WireProtocol(KNOWN_WIRESERVER_IP)
     protocol.mock_wire_data = WireProtocolData(mock_wire_data_file)
-    protocol.stop_mock_http_get = stop_mock_http_get
+    protocol.stop_mock_http_request = stop_mock_http_request
     protocol.stop_mock_crypt_util = stop_mock_crypt_util
+    protocol.stop = stop
```
Comment (narrieta) [inline tests/protocol/mock_wire_protocol.py]: those stop functions are mainly for debugging purposes; they come handy when there are conflicts with the mocks used by tests and the ones used here.

File: tests/protocol/test_hostplugin.py [hunk 5]
```diff
@@ -33,7 +33,7 @@
 from azurelinuxagent.common.utils import restutil
 from tests.protocol import mock_wire_protocol
 from tests.protocol.mockwiredata import WireProtocolData, DATA_FILE, DATA_FILE_NO_EXT
-from tests.protocol.test_wire import MockResponse
+from tests.protocol.test_wire import MockResponse as TestWireMockResponse
```
Comment (narrieta) [inline tests/protocol/test_hostplugin.py]: This one was so much fun to debug! This test uses 2 entities named MockResponse: the one imported here and another at the bottom of the file.

Which one gets used depends on whether patch is used as a decorator or as a function. Ah, the tangled web we weave!

The import alias takes care of this issue.

File: tests/protocol/test_hostplugin.py [hunk 6]
```diff
@@ -403,6 +403,7 @@ def test_validate_http_request(self):
 
         with mock_wire_protocol.create(DATA_FILE) as protocol:
             test_goal_state = protocol.client._goal_state
+            plugin = protocol.client.get_host_plugin()
```
Comment (narrieta) [inline tests/protocol/test_hostplugin.py]: Since now get_host_plugin() always fetches the goal state, this initialization was breaking the code below under the scope of the http_request mock.

I may come back (and also ask your help) to update more tests. Similar issues are all over the code.

File: tests/protocol/test_hostplugin.py [hunk 7]
```diff
@@ -564,151 +562,153 @@ def test_validate_get_extension_artifacts(self):
...
-            self.fail('IO error expected to be raised')
-        except IOError:
-            # expected
-            pass
-
-    @patch("azurelinuxagent.common.utils.restutil.http_get",
-           return_value=MockResponse(status_code=200, body=b''))
-    @patch("azurelinuxagent.common.protocol.healthservice.HealthService.report_host_plugin_versions")
-    def test_ensure_health_service_called(self, patch_http_get, patch_report_versions):
+        with patch("azurelinuxagent.common.utils.restutil.http_get") as patch_http_get:
```
Comment (narrieta) [inline tests/protocol/test_hostplugin.py]: I moved these patch decorators to be statements while I was debugging the issue with the mock request. Although this change is no longer needed I'm leaving there should we need to debug more.

File: tests/protocol/test_wire.py [hunk 8]
```diff
@@ -209,26 +209,28 @@ def test_get_host_ga_plugin(self, *args):
...
             self.assertEqual(goal_state.role_config_name, host_plugin.role_config_name)
 
-    @skip_if_predicate_true(lambda: True, "Needs to be re-enabled before release 2.2.47")
-    def test_upload_status_blob_default(self, *args):
-        """
-        Default status blob method is HostPlugin.
-        """
-        with create_mock_protocol(status_upload_blob=testurl, status_upload_blob_type=testtype) as protocol:
-            protocol.client.status_blob.vm_status = VMStatus(message="Ready", status="Ready")
+    def test_upload_status_blob_should_use_the_host_channel_by_default(self, *_):
```
Comment (narrieta) [inline tests/protocol/test_wire.py]: The changes below are the meat of this PR. Now we mock the interaction with the host/storage in terms of http requests, instead of reaching into implementation internals.

File: tests/protocol/test_wire.py [hunk 9]
```diff
@@ -516,227 +521,235 @@ def test_get_ext_conf_with_uri(self, *args):
...
-        with patch("azurelinuxagent.common.utils.restutil._http_request",
-                   side_effect=[mock_failed_response, mock_successful_response]):
-            with patch("azurelinuxagent.common.protocol.wire.WireClient.update_goal_state") as mock_update_goal_state:
-                with patch("azurelinuxagent.common.protocol.wire.WireClient.fetch", wraps=client.fetch) as patch_direct:
-                    with patch("azurelinuxagent.common.protocol.wire.WireClient.fetch_manifest_through_host",
-                               wraps=client.fetch_manifest_through_host) as patch_host:
-                        ret = client.fetch_manifest([VMAgentManifestUri(uri="uri1")])
-                        self.assertEquals(ret, "OK")
+        with mock_wire_protocol.create(mockwiredata.DATA_FILE) as protocol:
+            protocol.client.get_host_plugin()  # force initialization of the host plugin
```
Comment (pgombar) [inline tests/protocol/test_wire.py]: Why is this needed here and not in the test above?

File: tests/protocol/test_wire.py [hunk 10]
```diff
@@ -516,227 +521,235 @@ def test_get_ext_conf_with_uri(self, *args):
...
+            with patch("azurelinuxagent.common.utils.restutil.http_request", side_effect=http_request) as mock_request:
+                HostPluginProtocol.set_default_channel(False)
 
-        HostPluginProtocol.set_default_channel(False)
+                manifest = protocol.client.fetch_manifest([VMAgentManifestUri(uri=manifest_url)])
 
-        mock_failed_response = MockResponse(body=b"", status_code=httpclient.GONE)
-        mock_successful_response = MockResponse(body=b"OK", status_code=200)
+                self.assertEquals(manifest, manifest_xml)
+                self.assertTrue(any(args[0][0] == 'GET' and args[0][1] == manifest_url for args in mock_request.call_args_list))  # direct channel
```
Comment (pgombar) [inline tests/protocol/test_wire.py]: Don't we expect the first GET to succeed since we're mocking it? Why are we expecting several requests?

File: tests/protocol/test_wire.py (hunk 9 above)
Comment (narrieta) [reply tests/protocol/test_wire.py]: It is not needed, but simplifies the mock below. Initialization of the host plugin does a request for the goal state. Since the mock below is tracking goal state requests I would need to add code to filter out this one request. It is not too hard, but adding that code there would make the code less readable since it would distract from the actual intention of the test.

Some of the other tests don't track goal state requests so I do it only when it simplifies the code.

I'll add a coment.

File: tests/protocol/test_wire.py (hunk 10 above)
Comment (narrieta) [reply tests/protocol/test_wire.py]: Note that this test follows a different pattern than the other ones. This one was one of the first test I started with.

The underlying implementation of the protocol can trigger multiple requests while executing those fetch* and upload* methods. For example, if we get to use the host plugin we will trigger a request for the goal state, followed by puts to the health service for the plugin, etc.

The mocks in the other tests keep track of the interesting requests in http_request.urls and discard the rest. In this one test I am using mock_request.call_args_list instead. If you run the test under the debugger and set a breakpoint on the first assert you can check the call list and you will see a request for the goal state on top of the the GET for the manifest.

I think the pattern I used on the other tests allows for stronger assertions, and also makes tests a lot more readable so I'll change this one to follow that pattern as well.

Also, post-release, I'll see if I can capture that pattern in a more general tool that we can use for this kind of tests.

Comment (narrieta) [reply tests/protocol/test_wire.py]: I captured the pattern I was mentioning as mock_http_request in mocks.py

File: tests/common/test_event.py [hunk 11]
```diff
@@ -28,7 +28,8 @@
     WALAEventOperation, parse_xml_event, parse_json_event, AGENT_EVENT_FILE_EXTENSION, EVENTS_DIRECTORY
 from azurelinuxagent.common.future import ustr
 from azurelinuxagent.common.protocol.goal_state import GoalState
-from tests.protocol import mockwiredata, mock_wire_protocol
+from tests.protocol import mockwiredata
+from tests.protocol.mocks import mock_wire_protocol
```
Comment (narrieta) [inline tests/common/test_event.py]: I moved mock_wire_protocol to a different file (mocks.py)

Comment (narrieta) [reply tests/common/test_event.py]: I also changed the usage from

` with mock_wire_protocol.create(...)... 
`

to

` with mock_wire_protocol(...)... 
`

As I'm adding more functionality to the mocks the latter feels better.

File: tests/protocol/mock_wire_protocol.py [hunk 12]
```diff
@@ -1,84 +0,0 @@
...
-
-import contextlib
-import re
-from azurelinuxagent.common.protocol.wire import WireProtocol
-from azurelinuxagent.common.utils.restutil import KNOWN_WIRESERVER_IP, http_get
-from tests.tools import patch
-from tests.protocol.mockwiredata import WireProtocolData
-
-@contextlib.contextmanager
-def create(mock_wire_data_file):
```
Comment (narrieta) [inline tests/protocol/mock_wire_protocol.py]: moved this to mocks.py as mock_wire_protocol

File: tests/protocol/mocks.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+import contextlib
+import re
+from azurelinuxagent.common.protocol.wire import WireProtocol
+from azurelinuxagent.common.utils import restutil
+from tests.tools import patch
+from tests.protocol.mockwiredata import WireProtocolData
+
+
+@contextlib.contextmanager
+def mock_wire_protocol(mock_wire_data_file):
```
Comment (narrieta) [inline tests/protocol/mocks.py]: moved here from the previous mock_wire_protocol.py, with a few changes I am highlighting below

File: tests/protocol/mocks.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        # To minimize the impact of mocking restutil.http_request we only use the mock data for requests
+        # to the wireserver or requests starting with "mock-goal-state"
+        mock_data_re = re.compile(r'https?://(mock-goal-state|{0}).*'.format(restutil.KNOWN_WIRESERVER_IP.replace(r'.', r'\.')), re.IGNORECASE)
+
+        original_http_request = restutil.http_request
+
+        def http_request(method, url, data, headers=None, use_proxy=False, max_retry=restutil.DEFAULT_RETRIES, retry_codes=restutil.RETRY_CODES, retry_delay=restutil.DELAY_IN_SECONDS):
+            if method == 'GET' and mock_data_re.match(url) is not None:
+                return protocol.mock_wire_data.mock_http_get(url, headers, use_proxy, max_retry, retry_codes, retry_delay)
+            elif method == 'POST':
```
Comment (narrieta) [inline tests/protocol/mocks.py]: added support for POST

File: tests/protocol/mocks.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+            stop_mock_http_request.mock = None
+    stop_mock_http_request.mock = None
+
+    def stop_mock_crypt_util():
+        if stop_mock_crypt_util.mock is not None:
+            stop_mock_crypt_util.mock.stop()
+            stop_mock_crypt_util.mock = None
+    stop_mock_crypt_util.mock = None
+
+    def stop():
```
Comment (narrieta) [inline tests/protocol/mocks.py]: those stop functions are mainly for debugging purposes; they come handy when there are conflicts with the mocks used by tests and the ones used here.

File: tests/protocol/mocks.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        protocol.detect()
+
+        yield protocol
+
+    finally:
+        protocol.stop()
+
+
+@contextlib.contextmanager
+def mock_http_request(http_get_handler=None, http_post_handler=None, http_put_handler=None):
```
Comment (narrieta) [inline tests/protocol/mocks.py]: New mock useful for tests that need to interact with the host/storage at the HTTP request level

File: tests/common/test_event.py (hunk 11 above)
Comment (vrdmr) [reply tests/common/test_event.py]: Yup - Looks cleaner. Thanks

File: tests/protocol/test_wire.py [hunk 17]
```diff
@@ -516,227 +515,209 @@ def test_get_ext_conf_with_uri(self, *args):
...
-    def test_download_ext_handler_pkg_should_not_change_default_channel_if_host_fails(self, mock_get_artifact_request,
-                                                                                      *args):
-        mock_get_artifact_request.return_value = "dummy_url", "dummy_header"
-        protocol = WireProtocol("foo.bar")
-        HostPluginProtocol.set_default_channel(False)
+                urls = http_request.get_tracked_urls()
+                self.assertEquals(success, True, 'The download should have succeeded')
+                self.assertEquals(len(urls), 2, "Unexpected number of HTTP requests: [{0}]".format(urls))
+                self.assertEquals(urls[0], extension_url, "The first attempt should have been over the direct channel")
+                self.assertTrue(urls[1].endswith('/extensionArtifact'), "The retry attempt should have been over the host channel")
```
Comment (vrdmr) [inline tests/protocol/test_wire.py]: Do we need to validate the IP address to validate it went through the HGAP? I know the `/extensionArtifact` should be good enough to validate if the correct URI was hit.

Comment (narrieta) [reply tests/protocol/test_wire.py]: yes, that is one of the changes i want to do post-release

File: tests/protocol/test_wire.py (hunk 10 above)
Comment (pgombar) [reply tests/protocol/test_wire.py]: Got it, thanks for the update.
